### ISSUE TITLE
Conditional event binding

### ISSUE THREAD

**Issue by asiFarran** [vuejs/vue#7349]
### What problem does this feature solve?
Right now if we want an event handler to run conditionally we need to place the condition in the event handler which means that the event is still subscribed to and we pay the memory allocation fee for the event subscription (the underlying addEventListener and corresponding handler).

In some scenarios this is a pain. For example: say I've got mouse events (mouseover, mouseout) that are only meaningful on devices that actually have a pointer/mouse and so are meaningless on mobile and touch devices.

Right now, I'd have to create the event subscriptions and add the condition in the handler (or in this case no need even for that as the events would never fire) BUT I've still attached these handlers and allocated memory for them - which especially on memory constrained  platforms like mobile browsers is a waste.

By making the event subscription itself conditional we can avoid this.

### What does the proposed API look like?
In its simplest form we could just examine the handler provided and if falsy (or just null)  short-circuit the event subscription and NOT apply the underlying addEventListener operation.

This way the condition can appear in the event binding declaration itself:

```
<div @mouseover="condition ? handler : null" /> 
```

**Comment by sqal** [comment]
@asiFarran

> Right now, I'd have to create the event subscriptions and add the condition in the handler (or in this case no need even for that as the events would never fire) 

Actually there's another simple way you can handle this case. You can pass computed property with your listeners object (or null if condition is not met) to `v-on`, example: https://jsfiddle.net/c0Le92xe/

**Comment by Kingwl** [comment]
maybe you need  ?
`<div @mouseover="e => condition && handler(e)" />`

**Comment by yyx990803** [comment]
@Kingwl that still adds the listener, which is what OP wants to avoid.

@sqal 's suggestion is actually a valid workaround, and can be simplified to:

``` html
<div v-on="{ mouseover: condition ? handler : null }">
```

**Comment by asiFarran** [comment]
Yes, @sqal's solution is a good one and with inlining like @yyx990803's shows is close enough to what I had in mind. Thanks!  

This leads me to a deeper issue though: If a rebind happens (data change) and upon evaluation the condition changes (or more generally if the event spec object passed to v-on is different), existing event subscriptions are **not** cleaned up. They will still (hopefully) be disposed at the end of the component's lifecycle but not when they 'should' which is when v-on rebinds.

This may be an edge case that does not have a large impact for most scenarios but just for reference:

In my scenario I have a complex SVG and need to (optionally) attach mouse over/out handlers to specific elements based on some logic. 
The component is a long lived one and the underlying data changes leading to rebinding where I need to attach the event handlers to different elements at each rebind - hence my need to dispose of the previous ones so they don't dangle orphaned and sad. 
A typical solution - and what I'm going to end up doing - is to set up one 'top level' event listener (per event type) and let events bubble up to it  but to explain my motivation for trying it the Vue way first, there's quite a bit of processing that has to happen in my handlers which I had intended to pre-calculate  and bake right into the event subscriptions itself so that the handlers have less to do (quicker and more fluid response) and are not invoked at all for elements that don't require it (as opposed to filtering this out relying on DOM queries in the handler).

**Comment by VsevolodTrofimov** [comment]
@asiFarran 

how about

`<div @mouseover="handler">`

in methods
```javascript
handler: function(e) {
 if( ! this.condition) return
/* your handler */
}
```

**Comment by asiFarran** [comment]
@VsevolodTrofimov the whole point was to avoid the event subscription from happening at all. Applying the condition in the handler does not satisfy this requirement. Please note the comments above as they do provide a satisfactory resolution. 

I am closing this issue to reflect that.

**Comment by pmayer** [comment]
> `<div v-on="{ mouseover: condition ? handler : null }">`

Is it possible to pass `$event` and other arguments to handler?

**Comment by VsevolodTrofimov** [comment]
@pmayer 
>  `<div v-on="{ mouseover: condition ? $event => handler($event, arg) : null }">`

Or [curry](https://www.sitepoint.com/currying-in-functional-javascript/) the handler and use

 `<div v-on="{ mouseover: condition ? handler(arg) : null }">`

**Comment by brianwhu** [comment]
With the suggested solution

`<div v-on="{ mouseover: condition ? handler : null }">`

Is there a way to apply modifiers like ".stop.prevent"?

**Comment by sqal** [comment]
@brianwhu not in the template, just do it within handler function itself.

```
handler(event) {
  event.preventDefault();
  event.stopPropagation();
}
```

**Comment by DawidMyslak** [comment]
Nice one @pbastowski !

**Comment by cbo317110** [comment]
Thank you guys!

**Comment by vividvilla** [comment]
You may want to wrap with inline function if you are calling a handler with custom data. Something like this

`v-on="condition ? { mouseover: () => handler(somedata) } : {}"
`

**Comment by lk-jeffpeck** [comment]
This should be in Vue docs

**Comment by Artkoch** [comment]
> @DawidMyslak
> Just change it to the below and it will work
> 
> ```js
> v-on="condition ? { mouseover: handler } : {}"
> ```
> 
> or, if your handler is called `mouseover`
> 
> ```js
> v-on="condition ? { mouseover } : {}"
> ```

Is there a way to combine this this `.once`?

**Comment by posva** [comment]
You can use https://vuejs.org/v2/guide/render-function.html#Event-amp-Key-Modifiers

```vue
<button v-on="{ '~click': () => foo = new Date() }">Trigger only once</button>
```

**Comment by dietergeerts** [comment]
One question on this, will there be memory leaks or event  listeners not removed when I do the following, and the condition regularly changes?

```
<template>
<div  v-on="myListeners">
some content
</div>
</template>
<script>
...
computed: {
  myListeners() {
    return this.canExecute ? { click: () => this.$emit(...) } : {};
  },
},
...
</script>
```

So in short, does the `v-on` handles the change or not?

**Comment by yyx990803** [comment]
@dietergeerts it does.

**Comment by AlansCodeLog** [comment]
Does this work with native events? I couldn't get it working with `keydown.native`.

**Comment by Justineo** [comment]
FYI. Since 2.6 (not released yet) you will be able to apply a conditional event binding as follows:

```html
<div @[event]="handler" /> 
```

While `event` resolves to `null`, the binding will be removed.

**Comment by yyx990803** [comment]
@AlansCodeLog it should work with all modifiers. If it's not working you should open a new issue with a reproduction.

**Comment by AlansCodeLog** [comment]
I have opened an issue here: https://github.com/vuejs/vue/issues/9417

**Comment by capocasa** [comment]
> maybe you need ?
> `<div @mouseover="e => condition && handler(e)" />`

Browser JavaScript version

`<div @mouseover="condition && handler(arguments[0])" />`

**Comment by kieryk123** [comment]
Hey everyone, what about this solution, but in a functional component? It doesn't work for me...
`v-on="condition ? { mouseover: handler } : {}"`

How can I do this inside the functional component?

**Comment by pbastowski** [comment]
@kieryk123 Can you provide a CodeSandbox link to look at?

**Comment by arcreative** [comment]
In my case, I was trying to do this on a disabled button element (via Vuetify), but apparently Chrome doesn't fire the necessary events, and the element should be wrapped in a div or similar to catch the event.

**Comment by magistr4815** [comment]
Hello.
I tried to do it like this, but it does not work.
```js
// not working
v-on="{ [condition ? 'click.stop' : 'click'] : eventfunc }"
// or
// error
v-on="{ condition ? 'click.stop' : 'click' : eventfunc }"
```
You need to create a variable.

```js
// in created
this.blabla = condition ? 'click.stop' : 'click';

// in template
v-on="{ blabla : eventfunc }"
```
Is there any other solution?
Thanks in advance!

**Comment by proArtex** [comment]
Here is the trick for native events (using Vue v2.6.11):

`@mouseenter.native="condition && handler($event)"`

**Comment by AndrewBogdanovTSS** [comment]
`<div @[event]="handler" /> `
@Justineo was this syntax added to Vue?
I can't get it to work. Maybe you could send a link to docs where I can read more about it?

**Comment by aseidma** [comment]
@proArtex thank you, works for me as well!
`@keyup.delete="title.length === 0 && cancelTopicCreation`

Question is.... why?

**Comment by pbastowski** [comment]
@AndrewBogdanovTSS have a look at this fiddle https://jsfiddle.net/pbastowski/v0wt5qpo/23/

The syntax works just fine for normal and native clicks. Note that `.native` events are only available on Vue components and not plain HTML elements.

**Comment by AndrewBogdanovTSS** [comment]
@pbastowski simple string is working, yes, but what I need is to be able to resolve the value of event based on some reactive data, so I need something like
```
<h2 @[() => someProp ? 'mouseup' : null]="alert('Normal click on an HTML element')">H2 Element - click it</h2>
```
and such syntax doesn't work for me

**Comment by pbastowski** [comment]
@AndrewBogdanovTSS Just move your logic to a computed, see the updated example here  https://jsfiddle.net/pbastowski/v0wt5qpo/63/

**Comment by jhswart** [comment]
@AndrewBogdanovTSS I used this with the following format and it worked:

`@[click]="clickHandler"`

**Comment by coyotte508** [comment]
@AndrewBogdanovTSS another solution:

```
<h2 @[someProp&&`mouseup`]="alert('blablabla')">abcdef</h2>
```

Just make sure to not use spaces / quotes in the event name (why there's text in antiquotes)
